File: src/datastructures/MySinglyLinkedList.java
package datastructures;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.function.Consumer;

public class MySinglyLinkedList<T> implements Iterable<T> {
    
    public static class Node<T> {
        public T data;
        public Node<T> next;

        public Node(T data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node<T> head;
    private int size;

    public MySinglyLinkedList() {
        this.head = null;
        this.size = 0;
    }

    public void insertAtHead(T data) {
        Node<T> newNode = new Node<>(data);
        newNode.next = head;
        head = newNode;
        size++;
    }

    public void insertAtTail(T data) {
        Node<T> newNode = new Node<>(data);
        if (head == null) {
            head = newNode;
        } else {
            Node<T> curr = head;
            while (curr.next != null) {
                curr = curr.next;
            }
            curr.next = newNode;
        }
        size++;
    }

    public boolean remove(T data) {
        if (head == null) return false;

        if (head.data.equals(data)) {
            head = head.next;
            size--;
            return true;
        }

        Node<T> curr = head;
        while (curr.next != null) {
            if (curr.next.data.equals(data)) {
                curr.next = curr.next.next;
                size--;
                return true;
            }
            curr = curr.next;
        }
        return false;
    }

    public T find(T data) {
        Node<T> curr = head;
        while (curr != null) {
            if (curr.data.equals(data)) {
                return curr.data;
            }
            curr = curr.next;
        }
        return null;
    }

    public boolean contains(T data) {
        return find(data) != null;
    }

    public Object[] toArray() {
        Object[] arr = new Object[size];
        Node<T> curr = head;
        int idx = 0;
        while (curr != null) {
            arr[idx++] = curr.data;
            curr = curr.next;
        }
        return arr;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T removeHead() {
        if (head == null) return null;
        T data = head.data;
        head = head.next;
        size--;
        return data;
    }

    public T peekHead() {
        return head == null ? null : head.data;
    }

    public void clear() {
        head = null;
        size = 0;
    }

    @Override
    public void forEach(Consumer<? super T> action) {
        Node<T> curr = head;
        while (curr != null) {
            action.accept(curr.data);
            curr = curr.next;
        }
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private Node<T> current = head;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                T data = current.data;
                current = current.next;
                return data;
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("[");
        Node<T> curr = head;
        while (curr != null) {
            sb.append(curr.data);
            if (curr.next != null) {
                sb.append(", ");
            }
            curr = curr.next;
        }
        sb.append("]");
        return sb.toString();
    }
}

File: backend/src/dataStructures/MySinglyLinkedList.js
/**
 * MySinglyLinkedList — Custom Singly Linked List
 * CSD201 Data Structures Project
 * 
 * Used as the building block for adjacency lists in MyGraph
 * and as the internal storage for MyQueue.
 * 
 * Time Complexities:
 *   insertAtHead: O(1)
 *   insertAtTail: O(n)
 *   remove:       O(n)
 *   find:         O(n)
 *   contains:     O(n)
 *   size:         O(1) — cached
 *   toArray:      O(n)
 */

class SLLNode {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class MySinglyLinkedList {
  constructor() {
    this.head = null;
    this._size = 0;
  }

  /**
   * Insert a new node at the head of the list.
   * @param {*} data 
   * @returns {MySinglyLinkedList} this (for chaining)
   */
  insertAtHead(data) {
    const newNode = new SLLNode(data);
    newNode.next = this.head;
    this.head = newNode;
    this._size++;
    return this;
  }

  /**
   * Insert a new node at the tail of the list.
   * @param {*} data 
   * @returns {MySinglyLinkedList} this
   */
  insertAtTail(data) {
    const newNode = new SLLNode(data);
    if (!this.head) {
      this.head = newNode;
    } else {
      let current = this.head;
      while (current.next) {
        current = current.next;
      }
      current.next = newNode;
    }
    this._size++;
    return this;
  }

  /**
   * Remove the first node containing the given data.
   * @param {*} data 
   * @returns {boolean} true if removed, false if not found
   */
  remove(data) {
    if (!this.head) return false;

    if (this.head.data === data) {
      this.head = this.head.next;
      this._size--;
      return true;
    }

    let current = this.head;
    while (current.next) {
      if (current.next.data === data) {
        current.next = current.next.next;
        this._size--;
        return true;
      }
      current = current.next;
    }
    return false;
  }

  /**
   * Find the first node containing the given data.
   * @param {*} data 
   * @returns {SLLNode|null}
   */
  find(data) {
    let current = this.head;
    while (current) {
      if (current.data === data) return current;
      current = current.next;
    }
    return null;
  }

  /**
   * Check if the list contains the given data.
   * @param {*} data 
   * @returns {boolean}
   */
  contains(data) {
    return this.find(data) !== null;
  }

  /**
   * Convert the linked list to a plain array.
   * @returns {Array}
   */
  toArray() {
    const result = [];
    let current = this.head;
    while (current) {
      result.push(current.data);
      current = current.next;
    }
    return result;
  }

  /**
   * Execute a callback for each element.
   * @param {Function} callback (data, index)
   */
  forEach(callback) {
    let current = this.head;
    let index = 0;
    while (current) {
      callback(current.data, index);
      current = current.next;
      index++;
    }
  }

  /**
   * Get the number of elements.
   * @returns {number}
   */
  size() {
    return this._size;
  }

  /**
   * Check if the list is empty.
   * @returns {boolean}
   */
  isEmpty() {
    return this._size === 0;
  }

  /**
   * Clear all elements from the list.
   */
  clear() {
    this.head = null;
    this._size = 0;
  }
}

module.exports = MySinglyLinkedList;
